### PR TITLE
Fixed KWK_ASSERT 'interop' with BOOST_ENABLE_ASSERT_HANDLER.

### PR DESCRIPTION
**Pull request by psiha**
Binary/codegen size optimizations.

### CODE REVIEW

File: include/kwk/detail/assert.hpp [hunk 1]
```diff
@@ -16,30 +16,124 @@
...
+#   define KWK_ASSERT_TERMINATION()                                                                                            \
+        boost::assertion_failed_msg(condition, std::move(stream).str().c_str(), BOOST_CURRENT_FUNCTION, __FILE__, __LINE__);   \
+        KWK_UNREACHABLE()
+
+#else // Kiwaku assert implementation
+#   ifdef __GNUC__
+        [[ gnu::cold, gnu::noinline ]]
+#   endif
+        inline
+        auto && print_assertion_header(char const * const condition, char const * const file, unsigned const line) noexcept
```
Comment (jfalcou) [inline include/kwk/detail/assert.hpp:81]: auto& is enough I guess

File: include/kwk/detail/assert.hpp [hunk 2]
```diff
@@ -16,30 +16,124 @@
...
+
+    template < typename ArgTuple >
+#ifdef __GNUC__
+    [[ gnu::cold ]]
+#endif
+    [[ noreturn ]]
+    void assertion_handler(char const * const condition, char const * const file, unsigned const line, __::entupler<ArgTuple> const args) noexcept
+    {
+        decltype(auto) stream{ __::print_assertion_header(condition, file, line) };
+        std::apply
```
Comment (jfalcou) [inline include/kwk/detail/assert.hpp:111]: kumi::apply

File: include/kwk/detail/assert.hpp [hunk 3]
```diff
@@ -16,30 +16,124 @@
...
+    {                                                                              \
+      KIWAKU_ASSERT_DBG_BREAK();                                                   \
+      kwk::assertion_handler(#cond, __FILE__, __LINE__, __::args << __VA_ARGS__);  \
+    }                                                                              \
   } while(0) /**/
 
 #define KIWAKU_VERIFY(cond, ...) KIWAKU_ASSERT(cond, __VA_ARGS__)
 
-#endif
+#endif // KIWAKU_ASSERT
```
Comment (jfalcou) [inline include/kwk/detail/assert.hpp:139]: we don't do that :p

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/kwk/detail/assert.hpp [hunk 4]
```diff
@@ -16,30 +16,124 @@
...
+#   ifdef __GNUC__
+        [[ gnu::cold, gnu::noinline ]]
+#   endif
+        inline
+        auto print_assertion_header(char const * /*condition*/, char const * /*file*/, unsigned /*line*/) noexcept
+        {
+            return std::ostringstream{ "[KIWAKU] Assertion failed: " };
+        }
+
+#   ifdef __GNUC__
```
[the remark this answers is not in the document]
Comment (psiha) [reply include/kwk/detail/assert.hpp:68]: 👍

File: include/kwk/detail/assert.hpp (hunk 1 above)
Comment (psiha) [reply include/kwk/detail/assert.hpp]: 👍

File: include/kwk/detail/assert.hpp (hunk 3 above)
Comment (psiha) [reply include/kwk/detail/assert.hpp]: ok :P

**Review by jfalcou — APPROVED** [review]

**Comment by psiha** [comment]
> should we also change existing call to ASSERT to use , instead of << ?

@jfalcou just saw this question: well that would certainly simplify the cpp implementation portion of the assert macro implementation (i.e. lessen the compile time impact) but that would require more work on your part...

**Pull request merged**